cognite_toolkit/_builtin_modules/contextualization/p_and_id_parser/functions/contextualization_connection_writer/handler.py

```diff
@@ -0,0 +1,249 @@
+    # (max time is 3 days).
+    return {"status": status, "message": message}
+
+
+################# Data Classes #################
+
+# Logger using print
+class CogniteFunctionLogger:
+    def __init__(self, log_level: Literal["DEBUG", "INFO", "WARNING", "ERROR"] = "INFO"):
+        self.log_level = log_level.upper()
+
+    def debug(self, message: str):
+        if self.log_level == "DEBUG":
+            print(f"[DEBUG] {message}")
+
+    def info(self, message: str):
+        if self.log_level in ("DEBUG", "INFO"):
+            print(f"[INFO] {message}")
+
+    def warning(self, message: str):
+        if self.log_level in ("DEBUG", "INFO", "WARNING"):
+            print(f"[WARNING] {message}")
+
+    def error(self, message: str):
+        print(f"[ERROR] {message}")
+
+
+class ViewProperty(BaseModel, alias_generator=to_camel):
+    space: str
+    external_id: str
+    version: str
+    direct_relation_property: str | None = None
+    search_property: str = "name"
+
+    def as_view_id(self) -> dm.ViewId:
+        return dm.ViewId(space=self.space, external_id=self.external_id, version=self.version)
+
+
+class Mapping(BaseModel, alias_generator=to_camel):
+    file_source: ViewProperty
+    entity_source: ViewProperty
+
+
+
+class ConfigData(BaseModel, alias_generator=to_camel):
+    instance_spaces: list[str]
+    annotation_space: str
+    mappings: list[Mapping]
+
+
+class Config(BaseModel, alias_generator=to_camel):
+    data: ConfigData
+
+
+class State(BaseModel):
+    key: ClassVar[str] = FUNCTION_ID
+    last_cursor: str | None = None
+
+    @classmethod
+    def from_cdf(cls, client: CogniteClient) -> "State":
+        row = client.raw.rows.retrieve(db_name=RAW_DATABASE, table_name=RAW_TABLE, key=cls.key)
+        if row is None:
+            return cls()
+        return cls.model_validate(row.columns)
+
+    def to_cdf(self, client: CogniteClient) -> None:
+        client.raw.rows.insert(
+            db_name=RAW_DATABASE,
+            table_name=RAW_TABLE,
+            row=self._as_row(),
+        )
+
+    def _as_row(self) -> RowWrite:
+        return RowWrite(
+            key=self.key,
+            columns=self.model_dump(),
+        )
+
+
+
+################# Functions #################
+
+def execute(data: dict, client: CogniteClient) -> None:
+    logger = CogniteFunctionLogger(data.get("logLevel", "INFO")) # type: ignore[arg-type]
+    logger.debug("Starting connection write")
+    config = load_config(client, logger)
+    logger.debug("Loaded config successfully")
+
+    state = State.from_cdf(client)
+    connection_count = 0
+    for annotation_list in iterate_new_approved_annotations(state, client, config.data.annotation_space, logger):
+        annotation_by_source_by_node = to_direct_relations_by_source_by_node(annotation_list, config.data.mappings, logger)
+        connections = write_connections(annotation_by_source_by_node, client, logger)
+        connection_count += connections
+
+    state.to_cdf(client)
```

Comment:
How will number of rows accumulate over time? Is the number large enough to warrant a cleanup job, either here or separate, to take care of old cursors?

Reply:
This will always have only one row. While the last one will have one per file.